ui/controller/controller.go

```diff
@@ -516,3 +521,97 @@ func (c *Controller) SetTrackRatings(trackIDs []string, rating int) {
 		c.App.PlaybackManager.OnTrackRatingChanged(id, rating)
 	}
 }
+
+func (c *Controller) ShowDownloadDialog(tracks []*mediaprovider.Track) {
+	numTracks := len(tracks)
+	var fileName string
+	if numTracks == 1 {
+		parts := strings.Split(tracks[0].FilePath, "/")
+		fileName = parts[len(parts)-1]
+	} else {
+		fileName = "downloaded_tracks.zip"
+	}
+
+	dg := dialog.NewFileSave(
+		func(file fyne.URIWriteCloser, err error) {
+			if err != nil {
+				log.Println(err)
+				return
+			}
+
+			if file == nil {
+				return
+			}
+			if numTracks == 1 {
+				go c.downloadTrack(tracks[0], file.URI().Path())
+			} else {
+				go c.downloadTracks(tracks, file.URI().Path())
+			}
+
+		},
+		c.MainWindow)
+	dg.SetFileName(fileName)
+	dg.Show()
+}
+
+func (c *Controller) downloadTrack(track *mediaprovider.Track, filePath string) {
+	reader, err := c.App.ServerManager.Server.DownloadTrack(track.ID)
+	if err != nil {
+		log.Println(err)
+		return
+	}
+
+	file, err := os.Create(filePath)
+	if err != nil {
+		log.Println(err)
+		return
+	}
+	defer file.Close()
+
+	_, err = io.Copy(file, reader)
+	if err != nil {
+		log.Println(err)
+		return
+	}
+
+	log.Printf("Saved song %s to: %s\n", track.Name, filePath)
+}
+
+func (c *Controller) downloadTracks(tracks []*mediaprovider.Track, filePath string) {
+	zipFile, err := os.Create(filePath)
+	if err != nil {
+		log.Println(err)
+		return
+	}
+	defer zipFile.Close()
+
+	zipWriter := zip.NewWriter(zipFile)
+	defer zipWriter.Close()
+
+	for _, track := range tracks {
+		reader, err := c.App.ServerManager.Server.DownloadTrack(track.ID)
+		if err != nil {
+			log.Println(err)
+			continue
+		}
+
+		parts := strings.Split(track.FilePath, "/")
+		fileName := parts[len(parts)-1]
```

Comment:
you can use `fileName := filepath.Base(track.FilePath)` to get the name. (Add `path/filepath` to the file imports)